### ISSUE TITLE
Poor quality when rendering close up object in wild scenes

### ISSUE THREAD

**Issue by YZsZY** [nerfstudio-project/nerfstudio#1704]
Hi author, I recently tried to use nerfacto method to test on waymo dataset, I found that the hash encoding can easily reach 30+ psnr for far view in the test set, but the rendering quality is very bad for close view, for example, in the following case, the cylinder learns better from far view, but learns very bad in front view：
![001_22 420719146728516](https://user-images.githubusercontent.com/117158282/230343366-21dcd2a6-d5a4-4496-9294-7fd7839dfb4f.png)
![008_30 737947463989258](https://user-images.githubusercontent.com/117158282/230343428-7361c335-3e83-4c61-af52-e47871cfe0f8.png)

  Also the near pavement was badly learned. I noticed that nerfacto was equivalent to replacing the IPE encoding in mipnerf360 with hash encoding, so I tested the same data directly on mipnerf360. (The difference is only in the network, ngp uses hash encoding, mipnerf360 uses IPE encoding with a larger dimension MLP, other things including the bounding box size, the number of ray samples and the range of contraction are the same.) I found that mipnerf360 does not learn as well as ngp for far-away views, but learns very well for near-range views (left is mipnerf360, right is nerfacto)
![image](https://user-images.githubusercontent.com/117158282/230343942-2dbf51a7-d8f1-4f0b-9376-24ab51ccdfda.png)
    
  Here is a comparison of some other samples, you can see that compared to mipnerf360, nerfacto obviously learned badly in the near:(left is mipnerf360, right is nerfacto)
![image](https://user-images.githubusercontent.com/117158282/230345087-7d3021f8-64dd-4567-86bd-c972774b9f2c.png)

  I'm confused about this, it feels like nerfacto doesn't seem to have learned the geometry of the close up objects, yet the only difference between the two methods is the difference in encoding, and the sampling range is the same, I don't know what causes this phenomenon, I'd like to ask for some advice, thanks a lot!

**Comment by YZsZY** [comment]
You have some misunderstanding of picewise sampling, nerfacto this sampling method in the closer the more sampling, the more distant the less sampling, I think this situation should be too concentrated in the near field resulting in the corresponding hash feature grid can not effectively learn to the near features(a same grid needs to learn too many points in close proximity)

**Comment by YZsZY** [comment]
> I think this might be related to the Piecewise Sampler and Scene Contraction as stated in https://docs.nerf.studio/en/latest/nerfology/methods/nerfacto.html, the samping of nerfacto concentrates more on far objects, while NGP concentrate more on near objects.


You have some misunderstanding of picewise sampling, nerfacto this sampling method in the closer the more sampling, the more distant the less sampling, I think this situation should be too concentrated in the near field resulting in the corresponding hash feature grid can not effectively learn to the near features(a same grid needs to learn too many points in close proximity)

**Comment by donjiaking** [comment]
Hi, do you know how to solve this situation using nerfacto model? I ran into the same issue when using nuscenes dataset.

**Comment by YZsZY** [comment]
> Hi, do you know how to solve this situation using nerfacto model? I ran into the same issue when using nuscenes dataset.

Maybe this is a drawback when using ngp to encode a close-by object as nearby sampling is too concentrated resulting in the same voxel grid needs to learn too much, making the near field not very well learned. 

I tried to expand the translation matrix of the camera's pose to make the whole scene as large as possible, and the near voxel reduced the number of spatial points to be learned, and finally found that the quality of the near pose was effectively improved.

**Comment by donjiaking** [comment]
@YZsZY Thanks for the quick reply! How do you do it specifically? Could you share your code? I'm not very familiar with ngp hash encoding. Thanks!

**Comment by YZsZY** [comment]
> @YZsZY Thanks for the quick reply! How do you do it specifically? Could you share your code? I'm not very familiar with ngp hash encoding. Thanks!

I haven't had time to organize the code lately, but it's pretty simple. When you load the data, there is a step to scale normalize the poses, and nerfacto's operation is to limit the translation vector to a region of [-1,1], and you just need to adjust this scale.
In addition, I found that this scale normalization operation is useful for general 360 inward scenes, but for outdoor driving data sets such as outward views may lead to some other bugs, so you need to adjust the scale yourself

**Comment by donjiaking** [comment]
thans, I will try it

**Comment by HrsPythonix** [comment]
> > @YZsZY Thanks for the quick reply! How do you do it specifically? Could you share your code? I'm not very familiar with ngp hash encoding. Thanks!
> 
> I haven't had time to organize the code lately, but it's pretty simple. When you load the data, there is a step to scale normalize the poses, and nerfacto's operation is to limit the translation vector to a region of [-1,1], and you just need to adjust this scale. In addition, I found that this scale normalization operation is useful for general 360 inward scenes, but for outdoor driving data sets such as outward views may lead to some other bugs, so you need to adjust the scale yourself

I try to implement your idea, stated as below, please check if I miss something.

For example, by changing the scale factor and scene scale to 5.0 in dataparser like below, the poses and scene are scaled into [-5, 5]
```
        # Scale poses
        scale_factor = 5.0 #1.0 will scale to [-1, 1]
        if self.config.auto_scale_poses:
            scale_factor /= float(torch.max(torch.abs(poses[:, :3, 3])))
        scale_factor *= self.config.scale_factor

        poses[:, :3, 3] *= scale_factor

...
        self.config.scene_scale = 5.0
        # in x,y,z order
        # assumes that the scene is centered at the origin
        aabb_scale = self.config.scene_scale
        scene_box = SceneBox(
            aabb=torch.tensor(
                [[-aabb_scale, -aabb_scale, -aabb_scale], [aabb_scale, aabb_scale, aabb_scale]], dtype=torch.float32
            )
        )
...
```

One question is that if your pose and scene are not in [-1, 1], can the default enabled scene contraction be used? should we close it and use scenebox to normalize the positions?

**Comment by YZsZY** [comment]
> > > @YZsZY Thanks for the quick reply! How do you do it specifically? Could you share your code? I'm not very familiar with ngp hash encoding. Thanks!
> > 
> > 
> > I haven't had time to organize the code lately, but it's pretty simple. When you load the data, there is a step to scale normalize the poses, and nerfacto's operation is to limit the translation vector to a region of [-1,1], and you just need to adjust this scale. In addition, I found that this scale normalization operation is useful for general 360 inward scenes, but for outdoor driving data sets such as outward views may lead to some other bugs, so you need to adjust the scale yourself
> 
> I try to implement your idea, stated as below, please check if I miss something.
> 
> For example, by changing the scale factor and scene scale to 5.0 in dataparser like below, the poses and scene are scaled into [-5, 5]
> 
> ```
>         # Scale poses
>         scale_factor = 5.0 #1.0 will scale to [-1, 1]
>         if self.config.auto_scale_poses:
>             scale_factor /= float(torch.max(torch.abs(poses[:, :3, 3])))
>         scale_factor *= self.config.scale_factor
> 
>         poses[:, :3, 3] *= scale_factor
> 
> ...
>         self.config.scene_scale = 5.0
>         # in x,y,z order
>         # assumes that the scene is centered at the origin
>         aabb_scale = self.config.scene_scale
>         scene_box = SceneBox(
>             aabb=torch.tensor(
>                 [[-aabb_scale, -aabb_scale, -aabb_scale], [aabb_scale, aabb_scale, aabb_scale]], dtype=torch.float32
>             )
>         )
> ...
> ```
> 
> One question is that if your pose and scene are not in [-1, 1], can the default enabled scene contraction be used? should we close it and use scenebox to normalize the positions?

It still works, you compress the whole scene [-5,5] to [-2,2], and then normalize the space for hash encoding to [-1,1].

**Comment by YZsZY** [comment]
@HrsPythonix In addition, remember to wrap the entire scene of the scene box to adjust the size of the box, so that the box wrapped around your scaled pose, my personal advice is to adjust the scene box as large as possible for this outward perspective

**Comment by HrsPythonix** [comment]
> wrap

Do you mean that we should set scene_scale as large as possible? 

Also should we set the scale_factor to scale pose the same as scene_scale or some value less that the scene_scale?

**Comment by HrsPythonix** [comment]
> > @YZsZY Thanks for the quick reply! How do you do it specifically? Could you share your code? I'm not very familiar with ngp hash encoding. Thanks!
> 
> I haven't had time to organize the code lately, but it's pretty simple. When you load the data, there is a step to scale normalize the poses, and nerfacto's operation is to limit the translation vector to a region of [-1,1], and you just need to adjust this scale. In addition, I found that this scale normalization operation is useful for general 360 inward scenes, but for outdoor driving data sets such as outward views may lead to some other bugs, so you need to adjust the scale yourself

@YZsZY I review the related code these days, and I found several facts:
1. SceneBox is only activated when SceneContraction is disabled.
```
    def get_density(self, ray_samples: RaySamples) -> Tuple[TensorType, TensorType]:
        """Computes and returns the densities."""
        if self.spatial_distortion is not None:
            positions = ray_samples.frustums.get_positions()
            positions = self.spatial_distortion(positions)
            positions = (positions + 2.0) / 4.0
        else:
            positions = SceneBox.get_normalized_positions(ray_samples.frustums.get_positions(), self.aabb)
```
2. The default implementation of SceneContraction in nerfstudio will map all samples outside the [-1, 1] cube (which means [1, inf] or [-inf, -1])  to [1, 2] and [-2, -1], while keeping the samples inside [-1, 1] unchanged. Then the samples in [-2, 2] will be mapped to [0, 1] to act as inputs for tcnn.HashGrid, specific code as below:
```
scene_contraction = SceneContraction(order=float("inf"))
...
def contract(x):
            mag = torch.linalg.norm(x, ord=self.order, dim=-1)[..., None]
            return torch.where(mag < 1, x, (2 - (1 / mag)) * (x / mag))
......


if self.spatial_distortion is not None:
            positions = ray_samples.frustums.get_positions()
            positions = self.spatial_distortion(positions)
            positions = (positions + 2.0) / 4.0
        else:
            positions = SceneBox.get_normalized_positions(ray_samples.frustums.get_positions(), self.aabb)
```

So, by default, the samples inside [-1, 1] will occupy [0, 0.5] in HashGrid, all other point outside will occupy [0.5, 1]  in HashGrid, and the camera pose are all scaled into [-1, 1].

If we enlarge the scale factor for pose, for example to [-5, 5], which means camera poses are put into [-5, 5], I think this equals to putting more scene into the contacted part ([0.5, 1] in HashGrid), while putting less scene into the foreground part ([0, 0.5] in HashGrid), so the quality of foreground can become better. But the background will suffer.

**Comment by YZsZY** [comment]
@HrsPythonix You seem to have somehow not understood. The first problem about contract area is that the invariant area set in contract should just encompass the near field, which you need to try to set yourself.
In addition, the pose is scaled to ensure that the corresponding hash grid in the near field does not learn many points at once, so I expand the range so that the sampling points are more discrete.
You need to make sure that after you scale the pose, the invariant area you set still contains the near field, which is not necessarily [-1,1], but [-0.5,0.5], or [-5,5], anyhow, you need to try it yourself!

**Comment by YZsZY** [comment]
@HrsPythonix In fact, you only need to focus on the following points:
1. adjust your pose so that the near samples are not too concentrated on a hash grid, the best way is to zoom in on the pose, in which case your outermost SceneBox also needs to be scaled up.
2. After scaling the pose, adjust your contract area instead of using the original [-1,1], which is no longer guaranteed to contain the close-up area, and adjust the area so that it only contains the close-up area generated after the pose scale.

**Comment by YZsZY** [comment]
![image](https://user-images.githubusercontent.com/117158282/233545260-26ba1c39-cff6-403f-a28f-3d374919fcc4.png)
You can see that if the overall sampling range is small, it will be too concentrated in the near, which will lead to a hash grid that needs to interpolate too many points, however, there is a lot of detail in the near, resulting in failure to learn the details of the near, so the easiest way is to adjust the pose scale, so that the sampling range is expanded in scale, reducing the density of sampling in the near

**Comment by atztao** [comment]
Just used the uniform sample.

**Comment by YZsZY** [comment]
You think too easy....The use of log sampling itself is for more concentrated sampling in the near, the use of uniform sampling will make the near field sampling is not sufficient

**Comment by atztao** [comment]
Maybe we should custom ourself sampler.

**Comment by cxzhou35** [comment]
> Hi author, I recently tried to use nerfacto method to test on waymo dataset, I found that the hash encoding can easily reach 30+ psnr for far view in the test set, but the rendering quality is very bad for close view, for example, in the following case, the cylinder learns better from far view, but learns very bad in front view： ![001_22 420719146728516](https://user-images.githubusercontent.com/117158282/230343366-21dcd2a6-d5a4-4496-9294-7fd7839dfb4f.png) ![008_30 737947463989258](https://user-images.githubusercontent.com/117158282/230343428-7361c335-3e83-4c61-af52-e47871cfe0f8.png)
> 
> Also the near pavement was badly learned. I noticed that nerfacto was equivalent to replacing the IPE encoding in mipnerf360 with hash encoding, so I tested the same data directly on mipnerf360. (The difference is only in the network, ngp uses hash encoding, mipnerf360 uses IPE encoding with a larger dimension MLP, other things including the bounding box size, the number of ray samples and the range of contraction are the same.) I found that mipnerf360 does not learn as well as ngp for far-away views, but learns very well for near-range views (left is mipnerf360, right is nerfacto) ![image](https://user-images.githubusercontent.com/117158282/230343942-2dbf51a7-d8f1-4f0b-9376-24ab51ccdfda.png)
> 
> Here is a comparison of some other samples, you can see that compared to mipnerf360, nerfacto obviously learned badly in the near:(left is mipnerf360, right is nerfacto) ![image](https://user-images.githubusercontent.com/117158282/230345087-7d3021f8-64dd-4567-86bd-c972774b9f2c.png)
> 
> I'm confused about this, it feels like nerfacto doesn't seem to have learned the geometry of the close up objects, yet the only difference between the two methods is the difference in encoding, and the sampling range is the same, I don't know what causes this phenomenon, I'd like to ask for some advice, thanks a lot!

@YZsZY Thank you for sharing your experience, I am deal with a similar scene datasets, can you share your nerfstuio parameter settings or the training command? I use the nerfacto model to reconstruct my scene but the result are so bad.

**Comment by Jay-IPL** [comment]
> > Hi, do you know how to solve this situation using nerfacto model? I ran into the same issue when using nuscenes dataset.
> 
> Maybe this is a drawback when using ngp to encode a close-by object as nearby sampling is too concentrated resulting in the same voxel grid needs to learn too much, making the near field not very well learned.
> 
> I tried to expand the translation matrix of the camera's pose to make the whole scene as large as possible, and the near voxel reduced the number of spatial points to be learned, and finally found that the quality of the near pose was effectively improved.

may you let me know what are the scale_factor and scene_scale you used to get the improved results? their default values are both 1
